api/proquest/importer.py

```diff
@@ -0,0 +1,554 @@
+
+        if not token:
+            token = self._create_proquest_token(patron, configuration)
+
+        return token
+
+    def _get_book(self, patron, configuration, document_id):
+        """Get a book's content (in the case of open-access books) or a book's link otherwise.
+
+        :param patron: Patron object
+        :type patron: core.model.patron.Patron
+
+        :param configuration: Configuration object
+        :type configuration: ProQuestOPDS2ImporterConfiguration
+
+        :param document_id: ProQuest Doc ID
+        :type document_id: str
+
+        :return: Either an ACS link to the book or the book content
+        :rtype: Union[str, bytes]
+        """
+        token = self._get_or_create_proquest_token(patron, configuration)
+        iterations = 0
+
+        while True:
+            try:
+                book = self._api_client.get_book(self._db, token, document_id)
+
+                return book
+            except HTTPError as exception:
+                if exception.response.status_code != 401 or iterations >= 1:
+                    raise
+                else:
+                    token = self._create_proquest_token(patron, configuration)
+
+            iterations += 1
+
+    def extract_next_links(self, feed):
+        """Extracts "next" links from the feed.
+
+        :param feed: OPDS 2.0 feed
+        :type feed: Union[str, opds2_ast.OPDS2Feed]
+
+        :return: List of "next" links
+        :rtype: List[str]
+        """
+        return []
+
+    def patron_activity(self, patron, pin):
+        """Returns patron's loans
+
+        :param patron: A Patron object for the patron who wants to check out the book
+        :type patron: Patron
+
+        :param pin: The patron's alleged password
+        :type pin: string
+
+        :return: List of patron's loans
+        :rtype: List[LoanInfo]
+        """
+        now = datetime.datetime.utcnow()
+        loans = (
+            self._db.query(Loan)
+            .join(LicensePool)
+            .join(Collection)
+            .filter(
+                Collection.id == self._collection_id,
+                Loan.patron == patron,
+                or_(Loan.start is None, Loan.start <= now),
+                or_(Loan.end is None, Loan.end > now),
+            )
+        )
+
+        loan_info_objects = []
+
+        for loan in loans:
+            licensepool = get_one(self._db, LicensePool, id=loan.license_pool_id)
+
+            loan_info_objects.append(
+                LoanInfo(
+                    collection=self.collection,
+                    data_source_name=licensepool.data_source.name,
+                    identifier_type=licensepool.identifier.type,
+                    identifier=licensepool.identifier.identifier,
+                    start_date=loan.start,
+                    end_date=loan.end,
+                    fulfillment_info=None,
+                    external_identifier=None,
+                )
+            )
+
+        return loan_info_objects
```

Comment:
Yes, all the information about patron's loans/hold will be stored in Circulation Manager, ProQuest doesn't store information on their side. Unfortunately, `BaseCirculationAPI.patron_activity` doesn't have implementation of `patron_activity`, that's why I had to copy it from `LCPAPI`

Reply:
I don't like that we are making local loans look like remote loans, but I would be okay if you moved the implementation in `ODLAPI` to a superclass called something like `NoRemoteLoans`, and have both this class and `ODLAPI` subclass that class.

The implementation in `ODLAPI` is suitable for whenever the external API does not keep track of loans or holds, so it's the responsibility of the circulation manager to do things like hide loans that have expired and manage the holds queue.

Unfortunately this would mean moving a lot of code for managing a holds queue, and ProQuest doesn't need a holds queue, so it is a lot to ask. I'm OK with this implementation for now but please add a TODO to the docstring saying that this and code from ODLAPI should be refactored into a generic set of rules for any situation where the CM, not the remote API, is responsible for managing loans and holds.

Reply:
Sure, added a TODO to the docstring